imjoy/worker3.py

```diff
@@ -107,3 +113,77 @@
                 await result
         except Exception:  # pylint: disable=broad-except
             logger.error("Error in method %s: %s", job["num"], traceback.format_exc())
+
+
+class AsyncClient:
+    """Represent an async socketio client."""
+
+    def __init__(self, conn, opt, loop=None):
+        """Set up client instance."""
+        self.conn = conn
+        self.loop = loop or asyncio.get_event_loop()
+        self.opt = opt
+        self.queue = asyncio.Queue(loop=self.loop)
+        self.sio = socketio.AsyncClient()
+
+    def setup(self):
+        """Set up the plugin connection."""
+        logger.setLevel(logging.INFO)
+        if self.opt.debug:
+            logger.setLevel(logging.DEBUG)
+        self.sio.on("to_plugin_" + self.opt.secret, self.sio_plugin_message)
+
+        def on_disconnect():
+            if not self.opt.daemon:
+                self.conn.exit(1)
+
+        self.sio.on("disconnect", on_disconnect)
+
+    def connect(self):
+        """Connect to the socketio server."""
+        self.loop.run_until_complete(self.sio.connect(self.opt.server))
+        self.emit({"type": "initialized", "dedicatedThread": True})
+        logger.info("Plugin %s initialized", self.opt.id)
+        sys.stdout.flush()
+
+    def emit(self, msg):
+        """Emit a message to the socketio server."""
+        self.loop.create_task(self.sio.emit("from_plugin_" + self.opt.secret, msg))
+
+    async def sio_plugin_message(self, *args):
+        """Handle plugin message."""
+        data = args[0]
+        if data["type"] == "import":
+            self.emit({"type": "importSuccess", "url": data["url"]})
+        elif data["type"] == "disconnect":
+            self.conn.abort.set()
+            try:
+                if "exit" in self.conn.interface and callable(
+                    self.conn.interface["exit"]
+                ):
+                    self.conn.interface["exit"]()
+            except Exception as exc:  # pylint: disable=broad-except
+                logger.error("Error when exiting: %s", exc)
+            return args
+        elif data["type"] == "execute":
+            if not self.conn.executed:
+                await self.queue.put(data)
+            else:
+                logger.debug("Skip execution")
+                self.emit({"type": "executeSuccess"})
+        elif data["type"] == "message":
+            _data = data["data"]
+            await self.queue.put(_data)
+            logger.debug("Added task to the queue")
+        sys.stdout.flush()
+        return None
+
+    def wait_forever(self):
+        """Wait forever."""
+        wait_task = self.loop.create_task(self.sio.wait())
+        worker_tasks = [
+            task_worker(self.conn, self.queue, logger, self.conn.abort)
+            for i in range(10)
+        ]
+        self.loop.run_until_complete(asyncio.gather(*worker_tasks))
+        self.loop.run_until_complete(wait_task)
```

Comment:
So here socketio client loop is in the same loop as the task_workers, this will make the queue unnecessary, right? I think one potential issue here is that if the plugin running computation directly inside the main thread/loop, socketio won't have a chance to respond to the ping/pong signal, and then it will be disconnected.

Ideally, we would like to run the socketio in a separate thread, then use the queue to send tasks to the main thread/loop. 

(Main thread should be used to run the tasks to avoid potential issue for the plugin itself.)

Reply:
Slow computation shouldn't be run in the event loop. The event loop should never block or be slow. Should we schedule worker tasks on the executor thread pool?

Reply:
Also notice that `asyncio.Queue` is not thread-safe: https://docs.python.org/3/library/asyncio-queue.html#asyncio.Queue

Maybe an alternative solution is to use janus as we have before.

Reply:
Yes, in principle one should not run heavy computation inside async and they should use `loop.run_in_executor` -like solution. However, giving the fact that asyncio in python is relatively new, we do not enforce async programming in imjoy plugin development. The plugin functions can be synchronous, and we need to still call them in the loop. Even if all the plugin api functions are async, we cannot prevent the develop from executing heavy cpu tasks.

We cannot schedule worker tasks with a thread pool as well, because Tensorflow for example, do not allow a computational graph created in one thread to run in another, and there maybe other potential cross-thread issue. Since we don't know the dependencies of those worker tasks, we cannot really parallelized them into a thread pool.

For this part, we'd rather run the plugin api functions in the main thread, sequentially, and the developer make the thread pool themselves in needed.

Overall, I think the most feasible solution would be run the `sio.wait()` in a separate thread and use the queue to schedule worker tasks in the main thread (similar to our python2 worker).

Reply:
I think it makes more sense to have a dedicated worker thread for the worker queue tasks and then run the message handling in the main thread.

Then the message handling can be fast and the work tasks slow. The tasks will then run sequential and will be predictable.

Reply:
I afraid that won't work for all cases, the issue is that the plugin code is compiled and executed, api functions are also created in the main thread, if we send the plugin api functions to run in another thread, it will be problematic for tensorflow etc. 

We should let the plugin run in the main thread to avoid potential issues. And the plugin developers will also naturally assume that the plugin code are executed in main thread.  asyncio for example, it attach to the main thread by default. I am not sure what happened if you execute `import asyncio` in a different thread, whether a loop will be created.

By contrast, it will be much more controllable to just run the socketio in another thread.

Reply:
Yet another reason to run socketio in another thread is to support `from imjoy import api` in Jupyter notebook: https://github.com/oeway/ImJoy/issues/136#issuecomment-493690640

There, the plugin api functions are created in the main thread created with jupyter cells, which also means that the worker need to run in the main thread, then we have the socketio running in a background thread to hold the connection.

Reply:
I don't understand what you mean with api functions are created in the main thread. All plugin code are executed by the worker. So if the worker runs in a single thread, all the plugin code will be executed there. Am I missing something?